Fix initialization of LCD status line message

gudnimg opened the pull request:
Fixes #3581 

Change in memory:
Flash: -48 bytes
SRAM: -1 byte

gudnimg committed:
Initialise status line message in setup()

Fixes #3581

gudnimg committed:
Don't initialise lcd status message in definition

static variables are automatically zero initialised. Now that the
status line message is initialised in ultralcd_init(), we don't
need to set the variable in global scope.

Saves 22 bytes of flash and 1 byte of SRAM

leptun commented:
@gudnimg Do we even need this snippet of code anymore now that we always set the status line correctly? I think this code was used only for this welcome message and for old code which didn't set the status line with the specialized function.
```
   // Fill the rest of line to have nice and clean output
    for (uint8_t fillspace = 0; fillspace < LCD_WIDTH; fillspace++)
        if ((lcd_status_message[fillspace] <= 31))
            lcd_print(' ');
```

gudnimg commented:
> @gudnimg Do we even need this snippet of code anymore now that we always set the status line correctly? I think this code was used only for this welcome message and for old code which didn't set the status line with the specialized function.
> 
> ```
>    // Fill the rest of line to have nice and clean output
>     for (uint8_t fillspace = 0; fillspace < LCD_WIDTH; fillspace++)
>         if ((lcd_status_message[fillspace] <= 31))
>             lcd_print(' ');
> ```

I think you're right. Now `lcd_status_message` is only set in `lcd_updatestatus()` which always calls `lcd_finishstatus()`.

gudnimg committed:
Remove redundant for-loop

Now that lcd_status_message is now initialised correctly at boot-up,
this for-loop is no longer required. Now lcd_status_message is
only set in lcd_updatestatus() which always calls lcd_finishstatus()

lcd_finishstatus() makes sure the message does not exceed 20 characters

Saves 34 bytes of flash

leptun commented on Firmware/ultralcd.cpp:
```diff
@@ -94,7 +94,7 @@ static float manual_feedrate[] = MANUAL_FEEDRATE;
 static LongTimer lcd_status_message_timeout;
 static uint8_t lcd_status_message_level;
-static char lcd_status_message[LCD_WIDTH + 1] = WELCOME_MSG;
+static char lcd_status_message[LCD_WIDTH + 1];
```
I don't really understand how this saved 1B of RAM

gudnimg replied:
Perhaps a clue:
If we initialize the array with all elements as 1: it takes 1 extra byte of SRAM
`static char lcd_status_message[LCD_WIDTH + 1] = {1};`

If we initialize the array with all elements as 0: it takes no extra SRAM. 
`static char lcd_status_message[LCD_WIDTH + 1] = {0};`

leptun replied:
I think I know why this saved 1B of ram. It's because when you initialize the array with {1} it goes into the .data section, whch is in turn stored in flash. When you initialize the variable with {0} or don't initialize at all (in a global context), the variable gets placed in the .bss section which is initialized to 0x00 before jumping to main.
I see that the .data section is aligned to 2B while .bss is not. I think that's where a byte gets lost.

gudnimg replied:
That explanation sounds reasonable 👍
